core-plugins/src/main/java/io/cdap/plugin/batch/source/FTPBatchSource.java

```diff
@@ -181,15 +212,63 @@ public Schema getSchema() {
       return SCHEMA;
     }
 
+    public String extractPasswordFromUrl() {
+      int getLastIndexOfAtSign = path.lastIndexOf("@");
+      String authentication = path.substring(0, getLastIndexOfAtSign);
+      return authentication.substring(authentication.lastIndexOf(":") + 1);
+    }
+
+    public boolean authContainsSpecialCharacters() {
+      Pattern regularPasswordWithoutSpecialCharacters = Pattern.compile("[^A-Za-z0-9]");
+      Matcher regularPassword = regularPasswordWithoutSpecialCharacters.matcher(extractPasswordFromUrl());
+      return !regularPassword.matches();
+    }
+
     Map<String, String> getFileSystemProperties() {
-      if (fileSystemProperties == null) {
+      if (fileSystemProperties == null && !authContainsSpecialCharacters()) {
         return new HashMap<>();
       }
+      HashMap<String, String> fileSystemPropertiesMap = new HashMap<>();
+      if (fileSystemProperties != null) {
+        fileSystemPropertiesMap.putAll(GSON.fromJson(fileSystemProperties, MAP_STRING_STRING_TYPE));
+      }
+
       try {
-        return GSON.fromJson(fileSystemProperties, MAP_STRING_STRING_TYPE);
+        if (authContainsSpecialCharacters()) {
+          Path urlInfo;
+          String extractedPassword = extractPasswordFromUrl();
+          String encodedPassword = URLEncoder.encode(extractedPassword);
+          String validatePath = path.replace(extractedPassword, encodedPassword);
+          try {
+            urlInfo = new Path(validatePath);
+          } catch (Exception e) {
+            throw new IllegalArgumentException(String.format("Unable to parse url: %s %s", e.getMessage(), e));
+          }
+          int port = urlInfo.toUri().getPort();
+          String host = urlInfo.toUri().getAuthority().substring(urlInfo.toUri().getAuthority().lastIndexOf("@") + 1);
+          String user = urlInfo.toUri().getAuthority().split(":")[0];
+          if (urlInfo.toUri().getScheme().equals(FTP_PROTOCOL)) {
+            port = (port == -1) ? DEFAULT_FTP_PORT : port;
+            String cleanHostFTP = host.replace(":" + port, "");
```

Comment:
In case when urlInfo.toUri().getPort() returns -1 or undefined we add the default FTP/SFTP port. This is how it's required in FTP File System and we've followed the same pattern. 

This works as expected. After setting up the default port we are able to process what follows.